pkg/serializer/serializer.go

```diff
@@ -41,6 +41,8 @@ func (w *writer) do(v reflect.Value, sliceElem bool) {
 		w.doSlice(v)
 	case reflect.Struct:
 		w.doStruct(v, sliceElem)
+	case reflect.Map:
```

Comment:
How many map elements will be there?
Historically large map compilation was slow and made compiler to consume tremendous amounts of memory, and probably slows down start up. The issues was so bad that it become impossible to compile syzkaller on machine with any reasonable amount of RAM. So we switched to slices of structs everywhere.

Reply:
This particular map will be 5 element max, it is to simplify access to various bit fields when composing pseudo assembly instructions. I have a draft of that in https://github.com/aik/syzkaller/commits/ppc64 - the last patch.
btw that very last patch also adds a map with the instruction name as a key and recent PowerISA defines about 2500 but it is populated in runtime, is this still a problem?

Reply:
5 elements should be fine.

If a map is populated at runtime it's also not a subject to the compiler memory consumption problem. However, how is it populated? From what? Then there should be 2500 instructions in some other format, right? Where should I be looking at in that last patch?

Reply:
`pkg/ifuzz/ppc64/insns.go` is a list of instructions, it is big so the diff does not appear in github ui automatically. The map is built from the list. I was not quite sure if I want to have just a map for PPC64 and a list for x86 but since the common layer does not have a pointer to either anymore, I may revisit it. But if static maps are so bad in go, I am not sure what to do with it.

Reply:
I am not following.
Currently you have a static slice:
```
var insns_ppc = []*InsnPowerPC{
```
So this is good. That's what we generally use.

If you convert it to a map at runtime, it also won't create any issues with the compiler (b/c happens at runtime).
But do you need to convert it to a map at all? I would assume the main use is to select a random one, and it's easier with a slice. Or there is something special that you need to do?

p.s. it's not that static maps are bad in Go in general. Anything you can write manually is fine. It was only specific to auto-generated files where we generated tens/hundreds of thousands of elements. That's the reason we did not have support for maps in pkg/serializer so far.

Reply:
That something special is pseudo instructions (which are macros, really). Look here:
https://github.com/aik/syzkaller/commit/f19219d33da81045d341ec0fd49b80cac5bf2b2b#diff-a3bc25b4399ab84f8d1b404d2e106423R68

Reply:
I see. If you built that map lazily (I see some initOnce), then it should be fine.